datafusion/functions-nested/src/set_ops.rs

```diff
@@ -516,11 +516,18 @@ fn general_array_distinct<OffsetSize: OffsetSizeTrait>(
     let mut new_arrays = Vec::with_capacity(array.len());
     let converter = RowConverter::new(vec![SortField::new(dt)])?;
     // distinct for each list in ListArray
-    for arr in array.iter().flatten() {
+    for arr in array.iter() {
+        let last_offset: OffsetSize = offsets.last().copied().unwrap();
+        if arr.is_none() {
+            // Add same offset for null
+            offsets.push(last_offset);
+            continue;
+        }
+
+        let arr = arr.unwrap();
```

Comment:
I think another way to express this pattern without having to do `unwrap` is:

```suggestion
        let Some(arr) = arr else {
            // Add same offset for null
            offsets.push(last_offset);
            continue;
        }
```